src/Grouper.cpp

```diff
@@ -25,191 +25,283 @@
+      free(groups[i]);
+    }
+  }
+  if(groups)
+    free(groups);
 }
 
 /* *************************************** */
 
-bool Grouper::inGroup(Host *h) {
-  if(h == NULL || group_id_set == false)
-    return false;
+/**
+ * Returns group index. Calls newGroup() if no matching group is found.
+ *
+ * Returns -1 for unsupported sorting criteria, causes host to be skipped.
+ */
+int32_t Grouper::inGroup(Host *h) {
+  if(h == NULL)
+    return -1;
 
+  for(int32_t i = 0; i < numGroups; i++) {
   switch(sorter){
-  case column_asn:
-    return h->get_asn() == group_id_i;
-
-  case column_vlan:
-    return h->get_vlan_id() == group_id_i;
-
-  case column_local_network:
-  case column_local_network_id:
-    return h->get_local_network_id() == group_id_i;
-
-  case column_pool_id:
-    return h->get_host_pool() == group_id_i;
-
-  case column_mac:
-    return Utils::macaddr_int(h->get_mac()) == (u_int64_t)group_id_i;
-
-  case column_country:
-    {
-      char buf[32], *c = h->get_country(buf, sizeof(buf));
-      return (strcmp(group_id_s, c) == 0) ? true : false;
-    }
-    break;
-    
-  case column_os:
-    return h->get_os() ?
-      !strcmp(group_id_s, h->get_os()) :
-      !strcmp(group_id_s, (char*)UNKNOWN_OS);
+    case column_asn:
+      if(h->get_asn() == groups[i]->group_id_i)
+        return i;
+  
+    case column_vlan:
+      if(h->get_vlan_id() == groups[i]->group_id_i)
+        return i;
+  
+    case column_local_network:
+    case column_local_network_id:
+      if(h->get_local_network_id() == groups[i]->group_id_i)
+        return i;
+  
+    case column_pool_id:
+      if(h->get_host_pool() == groups[i]->group_id_i)
+        return i;
+  
+    case column_mac:
+      if(Utils::macaddr_int(h->get_mac()) == (u_int64_t)groups[i]->group_id_i)
+        return i;
+  
+    case column_country:
+      {
+        char buf[32], *c = h->get_country(buf, sizeof(buf));
+        if(groups[i]->group_id_s != NULL && strcmp(groups[i]->group_id_s, c) == 0)
+          return i;
+      }
+      break;
+      
+    case column_os:
+      if(h->get_os()) {
+        if(groups[i]->group_id_s != NULL && strcmp(groups[i]->group_id_s, h->get_os()) == 0)
+          return i;
+      } else {
+        if(groups[i]->group_id_s != NULL && strcmp(groups[i]->group_id_s, (char*)UNKNOWN_OS) == 0)
+          return i;
+      }
+  
+    default:
+      return -1;
+    };
+  }
 
-  default:
-    return false;
-  };
+  /* if reached no matching group found, create new one. */
+  return newGroup(h);
 }
 
 /* *************************************** */
 
-int8_t Grouper::newGroup(Host *h) {
+/**
+ * Creates a new group.
+ *
+ * returns group ID, -1 in case of errors or unsupported sorting criteria.
+ */
+int32_t Grouper::newGroup(Host *h) {
   char buf[32];
+  group **newg;
 
   if(h == NULL)
     return -1;
 
-  if(group_id_s){
-    free(group_id_s);
-    group_id_s = NULL;
+  newg = (group **)realloc(groups, sizeof(struct group) * (numGroups + 1));
```

Comment:
doing a realloc every time a new group is created could be an expensive operation

Reply:
Previous code iterated on the already sorted (but partial, for big sets) list, while my code grabs objects from the hash table as they come, unsorted. So my code has no way of knowing in advance how much memory to allocate for them.

I could make it reallocate in bunches of objects(ten at a time for example) instead of reallocating one by one.

Since the system malloc and friends usually already allocate in bunches bigger than the actually requested memory I thought leaving such preallocation work to them was more reasonable.